src/reformat_libyuv.c

```diff
@@ -77,6 +77,54 @@ avifResult avifImageRGBToYUVLibYUV(avifImage * image, const avifRGBImage * rgb)
     return AVIF_RESULT_NOT_IMPLEMENTED;
 }
 
+// Two-step replacement for AVIF_RGB_FORMAT_RGBA to 8-bit BT.601 full range YUV, which is missing from libyuv.
+static int avifABGRToJ420(const uint8_t * src_abgr,
+                          int src_stride_abgr,
+                          uint8_t * dst_y,
+                          int dst_stride_y,
+                          uint8_t * dst_u,
+                          int dst_stride_u,
+                          uint8_t * dst_v,
+                          int dst_stride_v,
+                          int width,
+                          int height)
+{
+    // A temporary buffer is needed to swap the R and B channels before calling ARGBToJ420().
```

Comment:
The API docs for libyuv are a little sparse, but I'd expect this conversion to work in place for most optimizations. The C code seems to handle that case. That might be something to clarify with @fbarchard.

Reply:
James: The `src_abgr` parameter points to a const buffer, so we can't modify the `src_abgr` buffer.

Yannis: I believe we just need to call `ARGBToJ420()` with the `dst_u` and `dst_v` buffers swapped. We use this trick elsewhere, as described by the following comment in rc/reformat_libyuv.c:

```
    // AVIF_RGB_FORMAT_RGB       *ToRGB24Matrix   matrixYVU
    // AVIF_RGB_FORMAT_BGR       *ToRGB24Matrix   matrixYUV
    // AVIF_RGB_FORMAT_RGB_565   *ToRGB565Matrix  matrixYUV
    // AVIF_RGB_FORMAT_BGRA      *ToARGBMatrix    matrixYUV
    // AVIF_RGB_FORMAT_RGBA      *ToARGBMatrix    matrixYVU
    // AVIF_RGB_FORMAT_ABGR      *ToRGBAMatrix    matrixYUV
    // AVIF_RGB_FORMAT_ARGB      *ToRGBAMatrix    matrixYVU
```

Reply:
> James: The `src_abgr` parameter points to a const buffer, so we can't modify the `src_abgr` buffer.

Sorry about that, I didn't look at the signature closely enough, I was just being hopeful I guess.

Reply:
Though the question still sta

> > James: The `src_abgr` parameter points to a const buffer, so we can't modify the `src_abgr` buffer.
> 
> Sorry about that, I didn't look at the signature closely enough, I was just being hopeful I guess.

Though the question still would be if it was ok to have the src and dst pointers be the same in this function and the calls to libyuv.

Reply:
> Yannis: I believe we just need to call `ARGBToJ420()` with the `dst_u` and `dst_v` buffers swapped. We use this trick elsewhere, as described by the following comment in rc/reformat_libyuv.c:

Yes, but the comment also contains:

```
    // In addition, swapping U and V in any of these calls, along with using the Yvu matrix instead of Yuv matrix,
    // swaps B and R in these orderings as well.
```

It is possible with `I420ToARGBMatrixFilter()` because it has a `const struct YuvConstants* yuvconstants` argument. However the YUV matrix coefficients are hard-coded in `ARGBToJ420()` and swapping `dst_u` and `dst_v` is not enough.
See Frank's comment: https://bugs.chromium.org/p/libyuv/issues/detail?id=936#c5

- - -

> Though the question still would be if it was ok to have the src and dst pointers be the same in this function and the calls to libyuv.

I do not follow.
Are you suggesting to store the temporary swapped RGB samples into `dst_y, dst_u, dst_v` to avoid an allocation, and then to call `ARGBToJ420()` in-place?
Or are you suggesting to modify the `src_abgr` samples in-place (they do not have to be const in the use case where `avifenc` reads a PNG and owns the samples)?

Anyway I hope that `ABGRToJ420()` will be available in libyuv at some point, and we will only need `avifABGRToJ420()` for earlier libyuv versions.

Reply:
`ARGBToABGR()` calls `ARGBShuffle()` which may call `ARGBShuffleRow_C()` which has the comment [`// To support in-place conversion.`](https://chromium.googlesource.com/libyuv/libyuv/+/9b17af9bef41aec80373a7d4689bb77bac33eab5/source/row_common.cc#3525). I did not check the intrinsic implementations but I would guess the whole function is in-place.

Reply:
Frank Barchard confirmed that `ARGBToABGR()` supports in-place conversion. In response to our question, he added a unit test for libyuv functions that should support in-place conversion:
https://chromium-review.googlesource.com/c/libyuv/libyuv/+/3828485